libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs

```diff
@@ -0,0 +1,154 @@
+              {
+                  var blobClient = storageAccount.CreateCloudBlobClient();
+                  var container = blobClient.GetContainerReference(containerName);
+                  container.CreateIfNotExistsAsync().Wait();
+                  return container;
+              });
+        }
+
+        /// <summary>
+        /// Get a blob name validated representation of an entity
+        /// </summary>
+        /// <param name="key">The key used to identify the entity</param>
+        /// <returns></returns>
+        private static string GetBlobName(string key)
+        {
+            if (string.IsNullOrEmpty(key)) throw new ArgumentNullException(nameof(key));
+
+            var blobName = HttpUtility.UrlEncode(key);
+            NameValidator.ValidateBlobName(blobName);
+            return blobName;
+        }
+
+        /// <summary>
+        /// Deletes entity blobs from the configured container
+        /// </summary>
+        /// <param name="keys">An array of entitiy keys</param>
+        /// <returns></returns>
+        public async Task Delete(string[] keys)
+        {
+            if (keys == null) throw new ArgumentNullException(nameof(keys));
+
+            foreach (var blobName in keys.Select(GetBlobName))
+            {
+                var blobReference = this.Container.Value.GetBlobReference(blobName);
+                await blobReference.DeleteAsync(
+                    DeleteSnapshotsOption.None, new AccessCondition { IfMatchETag = "*" }, new BlobRequestOptions(), new OperationContext()).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Retrieve entities from the configured blob container
+        /// </summary>
+        /// <param name="keys">An array of entitiy keys</param>
+        /// <returns></returns>
+        public async Task<StoreItems> Read(string[] keys)
+        {
+            if (keys == null) throw new ArgumentNullException(nameof(keys));
+
+            var storeItems = new StoreItems();
+            foreach (string blobName in keys.Select(GetBlobName))
+            {
+                using (var memoryStream = new MemoryStream())
+                {
+                    var blobReference = this.Container.Value.GetBlobReference(blobName);
+                    await blobReference.DownloadToStreamAsync(memoryStream).ConfigureAwait(false);
+                    using (var streamReader = new StreamReader(memoryStream, Encoding.UTF8))
+                    {
+                        var json = streamReader.ReadToEnd();
+                        var obj = JsonConvert.DeserializeObject(json, serializationSettings);
+                        IStoreItem storeItem = obj as IStoreItem;
+                        if (storeItem != null)
+                        {
+                            storeItem.eTag = blobReference.Properties.ETag;
+                        }
+
+                        storeItems[blobName] = storeItem;
+                    }

```

Comment:
Similar to my comment on `Delete`, I would think we'd want to process all the keys in parallel.